File: cursor_usage/pricing.py
"""Cursor model pricing rules (per million tokens, USD).

Primary source: https://cursor.com/docs/models-and-pricing
Provenance / confidence: cursor_usage.pricing_sources

Official doc has two pricing tables — do not conflate them:
- First-party models pool → Auto pricing + Composer pricing (#composer-pricing)
- API pool → Model pricing (GPT, Claude, Composer 1/2/2.5 base rows, etc.)

CSV slug composer-2.5-fast maps to Composer 2.5 (Fast) in the Composer pricing
table ($3 / $0.5 / $15), NOT the Composer 2.5 row in the API model table.

Monthly CSV samples validate totals but must not override documented rates.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping


@dataclass(frozen=True)
class ModelPricing:
    input: float
    cache_write: float
    cache_read: float
    output: float
    pool: str


# Included rows count against the paid plan allowance.
BILLABLE_KIND = "Included"

# Free rows represent free allowance usage.
# Strict mode: all Free rows count (USD in Cost or token formula).
# Official mode: Free with USD merges into Included; status-only Free is excluded.
FREE_KIND = "Free"
FREE_STATUS_ONLY_SKIP = "Free (status-only)"

# On-demand spend (pay-as-you-go beyond plan allowance).
ON_DEMAND_KIND = "On-demand"

ERRORED_KIND = "Errored, No Charge"
ABORTED_KIND = "Aborted, Not Charged"

SKIP_KINDS = frozenset({FREE_KIND, ERRORED_KIND, ABORTED_KIND})

# Canonical Kind values (case-insensitive match in normalize_kind).
_KIND_ALIASES: dict[str, str] = {
    "included": BILLABLE_KIND,
    "free": FREE_KIND,
    "on-demand": ON_DEMAND_KIND,
    "ondemand": ON_DEMAND_KIND,
    "errored, no charge": ERRORED_KIND,
    "aborted, not charged": ABORTED_KIND,
}


def normalize_kind(kind: str) -> str:
    """Return canonical Kind; matching is case-insensitive for known values."""
    key = (kind or "").strip().casefold()
    return _KIND_ALIASES.get(key, (kind or "").strip())


def is_billable_kind(kind: str) -> bool:
    return normalize_kind(kind) == BILLABLE_KIND


def is_free_kind(kind: str) -> bool:
    return normalize_kind(kind) == FREE_KIND


def is_on_demand_kind(kind: str) -> bool:
    return normalize_kind(kind) == ON_DEMAND_KIND


def is_skip_kind(kind: str) -> bool:
    return normalize_kind(kind) in SKIP_KINDS


# Max Mode billing (see docs/spec.md §2.4).
LONG_CONTEXT_INPUT_THRESHOLD = 272_000

CODEX_MAX_MODE_FAST_MODELS = frozenset({"gpt-5.3-codex", "gpt-5.3-codex-high"})
GPT_LONG_CONTEXT_MODELS = frozenset({"gpt-5.4-medium", "gpt-5.5-medium"})

CODEX_MAX_MODE_MULTIPLIER = 2.0
LONG_CONTEXT_INPUT_MULTIPLIER = 2.0
LONG_CONTEXT_OUTPUT_MULTIPLIER = 1.5


def parse_max_mode(value: str | None) -> bool:
    """True when CSV Max Mode column is Yes (case-insensitive)."""
    return (value or "").strip().casefold() == "yes"


def token_row_cost(
    pricing: ModelPricing,
    icw: int,
    icwo: int,
    cr: int,
    out: int,
    *,
    input_mult: float = 1.0,
    cache_write_mult: float | None = None,
    cache_read_mult: float | None = None,
    output_mult: float = 1.0,
) -> float:
    cwm = cache_write_mult if cache_write_mult is not None else input_mult
    crm = cache_read_mult if cache_read_mult is not None else input_mult
    return (
        icw / 1e6 * pricing.cache_write * cwm
        + icwo / 1e6 * pricing.input * input_mult
        + cr / 1e6 * pricing.cache_read * crm
        + out / 1e6 * pricing.output * output_mult
    )


def max_mode_adjusted_cost(
    model: str,
    pricing: ModelPricing,
    icw: int,
    icwo: int,
    cr: int,
    out: int,
    *,
    max_mode: bool,
) -> float | None:
    """Return Max Mode adjusted row cost, or None to use standard pricing."""
    if not max_mode:
        return None

    if model in CODEX_MAX_MODE_FAST_MODELS:
        m = CODEX_MAX_MODE_MULTIPLIER
        return token_row_cost(pricing, icw, icwo, cr, out, input_mult=m, output_mult=m)

    if model in GPT_LONG_CONTEXT_MODELS:
        if icw + icwo + cr <= LONG_CONTEXT_INPUT_THRESHOLD:
            return None
        return token_row_cost(
            pricing,
            icw,
            icwo,
            cr,
            out,
            input_mult=LONG_CONTEXT_INPUT_MULTIPLIER,
            cache_read_mult=LONG_CONTEXT_INPUT_MULTIPLIER,
            output_mult=LONG_CONTEXT_OUTPUT_MULTIPLIER,
        )

    return None


def parse_official_row_cost(value: str | None) -> float | None:
    """Parse Cost column when it holds a USD amount (not Included/Free status)."""
    if not value:
        return None
    text = value.strip()
    if not text or text in {BILLABLE_KIND, FREE_KIND, "-"}:
        return None
    try:
        return float(text)
    except ValueError:
        return None

AUTO_COMPOSER_MODELS = frozenset(
    {
        "auto",
        "composer-1",
        "composer-2",
        "composer-2-fast",
        "composer-2.5",
        "composer-2.5-fast",
    }
)
API_MODELS = frozenset(
    {
        "gpt-5.2",
        "gpt-5.2-codex",
        "gpt-5.3-codex",
        "gpt-5.3-codex-high",
        "gpt-5.4-medium",
        "gpt-5.5-medium",
        "claude-4.5-sonnet-thinking",
        "claude-4.6-sonnet-medium-thinking",
        "claude-4.6-opus-high-thinking",
        "claude-opus-4-7-thinking-high",
        "agent_review",
    }
)

# Bugbot agent_review: same token formula as Auto pool, optional discount (see RULE_SOURCES).
AGENT_REVIEW_DISCOUNT_RATIO = 1.0

PRICING: Mapping[str, ModelPricing] = {
    "auto": ModelPricing(
        input=1.25,
        cache_write=1.25,
        cache_read=0.25,
        output=6.0,
        pool="auto_composer",
    ),
    "composer-1": ModelPricing(
        input=1.25,
        cache_write=1.25,
        cache_read=0.125,
        output=10.0,
        pool="auto_composer",
    ),
    "composer-2": ModelPricing(
        input=0.5,
        cache_write=0.5,
        cache_read=0.2,
        output=2.5,
        pool="auto_composer",
    ),
    "composer-2-fast": ModelPricing(
        input=1.0,
        cache_write=1.0,
        cache_read=0.4,
        output=5.0,
        pool="auto_composer",
    ),
    # Composer pricing table → Composer 2.5 ($0.5 / $0.2 / $2.5); cache write at input rate.
    "composer-2.5": ModelPricing(
        input=0.5,
        cache_write=0.5,
        cache_read=0.2,
        output=2.5,
        pool="auto_composer",
    ),
    # Composer pricing table → Composer 2.5 (Fast); not the API-table Composer 2.5 row.
    "composer-2.5-fast": ModelPricing(
        input=3.0,
        cache_write=3.0,
        cache_read=0.5,
        output=15.0,
        pool="auto_composer",
    ),
    "gpt-5.2": ModelPricing(
        input=1.75,
        cache_write=1.75,
        cache_read=0.175,
        output=14.0,
        pool="api",
    ),
    "gpt-5.2-codex": ModelPricing(
        input=1.75,
        cache_write=1.75,
        cache_read=0.175,
        output=14.0,
        pool="api",
    ),
    "gpt-5.3-codex": ModelPricing(
        input=1.75,
        cache_write=1.75,
        cache_read=0.175,
        output=14.0,
        pool="api",
    ),
    "gpt-5.3-codex-high": ModelPricing(
        input=1.75,
        cache_write=1.75,
        cache_read=0.175,
        output=14.0,
        pool="api",
    ),
    "gpt-5.4-medium": ModelPricing(
        input=2.5,
        cache_write=2.5,
        cache_read=0.25,
        output=15.0,
        pool="api",
    ),
    "gpt-5.5-medium": ModelPricing(
        input=5.0,
        cache_write=5.0,
        cache_read=0.5,
        output=30.0,
        pool="api",
    ),
    "claude-4.5-sonnet-thinking": ModelPricing(
        input=3.0,
        cache_write=3.75,
        cache_read=0.3,
        output=15.0,
        pool="api",
    ),
    "claude-4.6-sonnet-medium-thinking": ModelPricing(
        input=3.0,
        cache_write=3.75,
        cache_read=0.3,
        output=15.0,
        pool="api",
    ),
    "claude-4.6-opus-high-thinking": ModelPricing(
        input=5.0,
        cache_write=6.25,
        cache_read=0.5,
        output=25.0,
        pool="api",
    ),
    "claude-opus-4-7-thinking-high": ModelPricing(
        input=5.0,
        cache_write=6.25,
        cache_read=0.5,
        output=25.0,
        pool="api",
    ),
    "agent_review": ModelPricing(
        input=1.25,
        cache_write=1.25,
        cache_read=0.25,
        output=6.0,
        pool="api",
    ),
}

File: web/src/lib/pricing.ts
export interface ModelPricing {
  input: number
  cacheWrite: number
  cacheRead: number
  output: number
  pool: string
}

export interface ResolvedModelPricing {
  requestedModel: string
  billingModel: string
  pricing: ModelPricing
  inferred: boolean
}

export const BILLABLE_KIND = 'Included'
export const FREE_KIND = 'Free'
export const FREE_STATUS_ONLY_SKIP = 'Free (status-only)'
export const ON_DEMAND_KIND = 'On-demand'

const KIND_ALIASES: Record<string, string> = {
  included: BILLABLE_KIND,
  free: FREE_KIND,
  'on-demand': ON_DEMAND_KIND,
  ondemand: ON_DEMAND_KIND,
  'errored, no charge': 'Errored, No Charge',
  'aborted, not charged': 'Aborted, Not Charged',
}

export function normalizeKind(kind: string): string {
  const key = (kind ?? '').trim().toLowerCase()
  return KIND_ALIASES[key] ?? (kind ?? '').trim()
}

export function isBillableKind(kind: string): boolean {
  return normalizeKind(kind) === BILLABLE_KIND
}

export function isFreeKind(kind: string): boolean {
  return normalizeKind(kind) === FREE_KIND
}

export function isOnDemandKind(kind: string): boolean {
  return normalizeKind(kind) === ON_DEMAND_KIND
}

export const LONG_CONTEXT_INPUT_THRESHOLD = 272_000
export const CODEX_MAX_MODE_FAST_MODELS = new Set(['gpt-5.3-codex'])
export const GPT_LONG_CONTEXT_MODELS = new Set(['gpt-5.4', 'gpt-5.5'])
export const CODEX_MAX_MODE_MULTIPLIER = 2.0
export const LONG_CONTEXT_INPUT_MULTIPLIER = 2.0
export const LONG_CONTEXT_OUTPUT_MULTIPLIER = 1.5
export const AGENT_REVIEW_DISCOUNT_RATIO = 1.0

const MODEL_ALIASES: Record<string, string> = {
  'cursor-grok-4.5': 'grok-4.5',
}

const MODEL_NORMALIZATION_RULES: Array<{ prefix: string; suffixes: readonly string[] }> = [
  { prefix: 'gpt-5.', suffixes: ['-xhigh', '-high', '-medium'] },
  { prefix: 'claude-', suffixes: ['-medium-thinking', '-high-thinking', '-thinking-high', '-thinking', '-medium'] },
  { prefix: 'grok-4.5', suffixes: ['-xhigh', '-high'] },
  { prefix: 'cursor-grok-4.5', suffixes: ['-high', '-xhigh'] },
]

export function parseMaxMode(value: string | undefined | null): boolean {
  return (value ?? '').trim().toLowerCase() === 'yes'
}

export interface TokenCostBreakdown {
  icw: number
  icwo: number
  cacheRead: number
  output: number
}

export function tokenRowCostComponents(
  pricing: ModelPricing,
  icw: number,
  icwo: number,
  cr: number,
  out: number,
  opts: {
    inputMult?: number
    cacheWriteMult?: number
    cacheReadMult?: number
    outputMult?: number
  } = {},
): TokenCostBreakdown {
  const inputMult = opts.inputMult ?? 1
  const cwm = opts.cacheWriteMult ?? inputMult
  const crm = opts.cacheReadMult ?? inputMult
  const outputMult = opts.outputMult ?? 1
  return {
    icw: (icw / 1e6) * pricing.cacheWrite * cwm,
    icwo: (icwo / 1e6) * pricing.input * inputMult,
    cacheRead: (cr / 1e6) * pricing.cacheRead * crm,
    output: (out / 1e6) * pricing.output * outputMult,
  }
}

export function tokenRowCost(
  pricing: ModelPricing,
  icw: number,
  icwo: number,
  cr: number,
  out: number,
  opts: {
    inputMult?: number
    cacheWriteMult?: number
    cacheReadMult?: number
    outputMult?: number
  } = {},
): number {
  const parts = tokenRowCostComponents(pricing, icw, icwo, cr, out, opts)
  return parts.icw + parts.icwo + parts.cacheRead + parts.output
}

export function maxModeAdjustedCost(
  model: string,
  pricing: ModelPricing,
  icw: number,
  icwo: number,
  cr: number,
  out: number,
  maxMode: boolean,
): number | null {
  if (!maxMode) return null

  if (CODEX_MAX_MODE_FAST_MODELS.has(model)) {
    const m = CODEX_MAX_MODE_MULTIPLIER
    return tokenRowCost(pricing, icw, icwo, cr, out, { inputMult: m, outputMult: m })
  }

  if (GPT_LONG_CONTEXT_MODELS.has(model)) {
    if (icw + icwo + cr <= LONG_CONTEXT_INPUT_THRESHOLD) return null
    return tokenRowCost(pricing, icw, icwo, cr, out, {
      inputMult: LONG_CONTEXT_INPUT_MULTIPLIER,
      cacheReadMult: LONG_CONTEXT_INPUT_MULTIPLIER,
      outputMult: LONG_CONTEXT_OUTPUT_MULTIPLIER,
    })
  }

  return null
}

export function parseOfficialRowCost(value: string | undefined | null): number | null {
  if (!value) return null
  const text = value.trim()
  if (!text || text === BILLABLE_KIND || text === FREE_KIND || text === '-') return null
  const n = Number(text)
  return Number.isFinite(n) ? n : null
}

export const PRICING: Record<string, ModelPricing> = {
  auto: { input: 1.25, cacheWrite: 1.25, cacheRead: 0.25, output: 6.0, pool: 'auto_composer' },
  'composer-1': { input: 1.25, cacheWrite: 1.25, cacheRead: 0.125, output: 10.0, pool: 'auto_composer' },
  'composer-2': { input: 0.5, cacheWrite: 0.5, cacheRead: 0.2, output: 2.5, pool: 'auto_composer' },
  'composer-2-fast': { input: 1.0, cacheWrite: 1.0, cacheRead: 0.4, output: 5.0, pool: 'auto_composer' },
  'composer-2.5': { input: 0.5, cacheWrite: 0.5, cacheRead: 0.2, output: 2.5, pool: 'auto_composer' },
  'composer-2.5-fast': { input: 3.0, cacheWrite: 3.0, cacheRead: 0.5, output: 15.0, pool: 'auto_composer' },
  'gpt-5.2': { input: 1.75, cacheWrite: 1.75, cacheRead: 0.175, output: 14.0, pool: 'api' },
  'gpt-5.2-codex': { input: 1.75, cacheWrite: 1.75, cacheRead: 0.175, output: 14.0, pool: 'api' },
  'gpt-5.3-codex': { input: 1.75, cacheWrite: 1.75, cacheRead: 0.175, output: 14.0, pool: 'api' },
  'gpt-5.4': { input: 2.5, cacheWrite: 2.5, cacheRead: 0.25, output: 15.0, pool: 'api' },
  'gpt-5.5': { input: 5.0, cacheWrite: 5.0, cacheRead: 0.5, output: 30.0, pool: 'api' },
  'claude-4.5-sonnet': { input: 3.0, cacheWrite: 3.75, cacheRead: 0.3, output: 15.0, pool: 'api' },
  'claude-4.6-sonnet': { input: 3.0, cacheWrite: 3.75, cacheRead: 0.3, output: 15.0, pool: 'api' },
  'claude-4.6-opus': { input: 5.0, cacheWrite: 6.25, cacheRead: 0.5, output: 25.0, pool: 'api' },
  'claude-opus-4-7': { input: 5.0, cacheWrite: 6.25, cacheRead: 0.5, output: 25.0, pool: 'api' },
  'gpt-5.6-sol': { input: 5.0, cacheWrite: 6.25, cacheRead: 0.5, output: 30.0, pool: 'api' },
  'grok-4.5': { input: 2.0, cacheWrite: 2.0, cacheRead: 0.5, output: 6.0, pool: 'auto_composer' },
  agent_review: { input: 1.25, cacheWrite: 1.25, cacheRead: 0.25, output: 6.0, pool: 'api' },
}

function applyModelAliases(model: string): string {
  return MODEL_ALIASES[model] ?? model
}

export function normalizeModel(model: string): string {
  if (model in PRICING) return model

  for (const rule of MODEL_NORMALIZATION_RULES) {
    if (!model.startsWith(rule.prefix)) continue
    for (const suffix of rule.suffixes) {
      if (model.endsWith(suffix) && model.length > suffix.length) {
        return applyModelAliases(model.slice(0, -suffix.length))
      }
    }
  }

  return applyModelAliases(model)
}

export function resolveModelPricing(model: string): ResolvedModelPricing | null {
  const billingModel = normalizeModel(model)
  const pricing = PRICING[billingModel]
  if (pricing) {
    return {
      requestedModel: model,
      billingModel,
      pricing,
      inferred: billingModel !== model,
    }
  }
  return null
}

export function rowCostBreakdown(
  model: string,
  icw: number,
  icwo: number,
  cr: number,
  out: number,
  maxMode = false,
): TokenCostBreakdown {
  if (model === 'agent_review') {
    const parts = tokenRowCostComponents(PRICING.auto, icw, icwo, cr, out)
    const ratio = AGENT_REVIEW_DISCOUNT_RATIO
    return {
      icw: parts.icw * ratio,
      icwo: parts.icwo * ratio,
      cacheRead: parts.cacheRead * ratio,
      output: parts.output * ratio,
    }
  }

  const resolved = resolveModelPricing(model)
  if (!resolved) return { icw: 0, icwo: 0, cacheRead: 0, output: 0 }
  const { billingModel, pricing } = resolved

  if (maxMode && CODEX_MAX_MODE_FAST_MODELS.has(billingModel)) {
    const m = CODEX_MAX_MODE_MULTIPLIER
    return tokenRowCostComponents(pricing, icw, icwo, cr, out, { inputMult: m, outputMult: m })
  }

  if (maxMode && GPT_LONG_CONTEXT_MODELS.has(billingModel) && icw + icwo + cr > LONG_CONTEXT_INPUT_THRESHOLD) {
    return tokenRowCostComponents(pricing, icw, icwo, cr, out, {
      inputMult: LONG_CONTEXT_INPUT_MULTIPLIER,
      cacheReadMult: LONG_CONTEXT_INPUT_MULTIPLIER,
      outputMult: LONG_CONTEXT_OUTPUT_MULTIPLIER,
    })
  }

  return tokenRowCostComponents(pricing, icw, icwo, cr, out)
}

export function rowCost(
  model: string,
  icw: number,
  icwo: number,
  cr: number,
  out: number,
  maxMode = false,
): number {
  if (model === 'agent_review') {
    const p = PRICING.auto
    return tokenRowCost(p, icw, icwo, cr, out) * AGENT_REVIEW_DISCOUNT_RATIO
  }
  const resolved = resolveModelPricing(model)
  if (!resolved) return 0
  const { billingModel, pricing } = resolved
  const adjusted = maxModeAdjustedCost(billingModel, pricing, icw, icwo, cr, out, maxMode)
  if (adjusted !== null) return adjusted
  return tokenRowCost(pricing, icw, icwo, cr, out)
}

export function resolveRowCost(
  rowCostValue: string | undefined | null,
  model: string,
  icw: number,
  icwo: number,
  cr: number,
  out: number,
  maxMode = false,
): number {
  const annotated = parseOfficialRowCost(rowCostValue)
  if (annotated !== null) return annotated
  return rowCost(model, icw, icwo, cr, out, maxMode)
}
